design/read-only-registry.md

```diff
@@ -0,0 +1,76 @@
+# Support for Read-Only Registries
+
+## Goals
+
+- Configure kotsadm to use application images from a private registry
+- Support using read-only registry during initial installation
+- Support using read-only registry on application updates
+- Support using read-only registry on when settings are modified on the Registry Settings page in Admin Console
+- Support readon-only registries in airgap and on-line modes
+
+## Non Goals
+
+- Delivering images to the read-only registry.
+
+## Background
+
+Some production enviroments restrict image registry use to read-only mode.  This is an environment wide restriction.  Images are side-loaded into the registry from a trusted internal source.  Right now kotsadm will always push images to the private registry, making it impossible to use in such an environment.
+
+## Detailed Design
+
+### Install
+
+A new argument will be added to `kots install` command to flag the registry as a read-only registry.
+
+```
+kubectl kots install --kotsadm-registry my.private.registry/myapp \
+  --registry-username pulluser \
+  --registry-password pullpassword \
+  --registry-is-readonly \
+  app-slug
+```
+
+When the above command is used, the following will happen:
+
+1. All kotsadm pods will be rewritten to pull images from the private registry. This is existing functionality.
+1. Provided registry credentials will be saved in `.dockerconfigjson` secret in the app's namespace.
+1. During license upload in Admin Console UI, registry settings will be populated with these values.
+1. Afte the application is installed, Registry Settings page will be populated with these settings.
+
+Note: this changes current functionality where registry information specified in the command line arguments is not saved and is not applied to the application pods.
+
+### Update
+
+Online install will download updates, but will not push images to the private registry as long as the read-only flag is set.
+
+Aigapped installs will accept airgap bundle, but will not push images to the private registry as long as the read-only flag is set.
+
+Stretch: Replicated could provide imageless airgap bundles.
+
+### Registry Settings page
+
+1. Registry settings page will have a checkbox named Read only.
+1. In aigapped environments this is always checked and disabled and all other fields can be modified.
```

Comment:
Why disable this in airgap? Do you mean inn kurl? If this is kurl, wouldn't it always be unchecked?

Reply:
We can't push images in airgapped mode with current UI design. These settings are disabled right now.

Reply:
I'm confused. Doesn't uploading an airgap package currently result in kotsadm pushing images to the configured registry?

Reply:
@laverya this is about the Registry Settings page

Reply:
@divolgin is the intent to disallow changing from RO to RW (or RW to RO) on airgap, but still allow either to be set depending on what was chosen during initial setup?

Reply:
yeah...   this is actually wrong.  it should be "always disabled" but not "always checked"